### ISSUE TITLE
What is 'will be cast to bool' mean?

### ISSUE THREAD

**Issue by swpper** [keras-team/keras#16022]
Hello, I am using tf.keras.metrics and confuse with the annotation for y_true in the source code:
```
def update_state(self, y_true, y_pred, sample_weight=None):
    """Accumulates true positive and false positive statistics.
    Args:
      y_true: The ground truth values, with the same dimensions as `y_pred`.
        Will be cast to `bool`.
      y_pred: The predicted values. Each element must be in the range `[0, 1]`.
      sample_weight: Optional weighting of each example. Defaults to 1. Can be a
        `Tensor` whose rank is either 0, or the same rank as `y_true`, and must
        be broadcastable to `y_true`.
```
What is the meaning of Will be cast to bool? I want to see the precision of each step, does it means that I can use it on a float regression model output(0~1) and it will change it to 0 or 1 by threshold both for y_true and y_pred? But the demo below shows a different case. There seems exists a hidden rule/logic which leads to different results on condition of y_true>y_pred and y_true<y_pred even though they both larger than threshold. What is the logic of the metrics based on  `confusion matrix`? Or I should not use it on regression model?
```
import tensorflow as tf

y_true = [0, 1, 0.6, 0]
y_pre =  [1, 1, 0.3, 1]

ms_name = ['TruePositives', 'TrueNegative', 'FalsePositives', 'FalseNegative']
ms = [   
    tf.keras.metrics.TruePositives(), tf.keras.metrics.TrueNegatives(),
    tf.keras.metrics.FalsePositives(), tf.keras.metrics.FalseNegatives()
]
for m_name, m in zip(ms_name, ms):
    m.update_state(y_true, y_pre)
    res = m.result().numpy()
    del m
    print('\n', m_name, res)


metric_name = ['precision']
metrics = [tf.keras.metrics.Precision(0.1)]

for m_name, m in zip(metric_name, metrics):
    m.update_state(y_true, y_pre)
    res = m.result().numpy()
    del m
    print('\n', m_name, res)
```

**Comment by sachinprasadhs** [comment]
Could you please link to the code which you have mentioned, When checking the metrics update_state for TruePositives I see below details https://github.com/keras-team/keras/blob/d8fcb9d4d4dad45080ecfdd575483653028f8eda/keras/metrics.py#L1060-L1072

**Comment by swpper** [comment]
Hello @sachinprasadhs 
What I mentioned is below at `tf.keras.metrics.Precision`

https://github.com/keras-team/keras/blob/d8fcb9d4d4dad45080ecfdd575483653028f8eda/keras/metrics.py#L1394

Lately, I learned that the metrics like `Truepositives` directly treat the values in `y_true` as category labels. But the metrics calculated based on `truepositive` and the other three like `Precision` will change `y_true` to category labels by threshold along with `y_pred`. Do I understand right?

**Comment by swpper** [comment]
@sachinprasadhs  Sorry, I update the code link in master where I want to asked for `tf.keras.metrics.Precision`.

https://github.com/keras-team/keras/blob/8ecef127f70db723c158dbe9ed3268b3d610ab55/keras/metrics/metrics.py#L784-L809

What I want to figure out is  

> Lately, I learned that the metrics like `Truepositives` directly treat the values in `y_true` as category labels. But the metrics calculated based on `truepositive` and the other three like `Precision` will change `y_true` to category labels by threshold as the same to `y_pred`. Do I understand right?

**Comment by sachinprasadhs** [comment]
Yes, you need to transform `y_true` into 0/1, but `y_pred` can be any value between [0,1]
